AK/InsertionSort.h

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2022, Marc Luqué <marc.luque@outlook.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/StdLibExtras.h>
+
+namespace AK {
+
+/* Standard Insertion Sort.
+ * NOTE: `end` is inclusive.
+ */
+template<typename Collection, typename Compare>
+void insertion_sort(Collection& col, int start, int end, Compare compare)
+{
+    for (int i = start; i <= end; ++i) {
+        for (int j = i; j > 0 && compare(col[j], col[j - 1]); --j) {
+            swap(col[j], col[j - 1]);
+        }
+    }
+}
+
+// NOTE: `end` is exclusive.
+template<typename Iterator, typename Compare>
+void insertion_sort(Iterator start, Iterator end, Compare compare)
+{
+    for (Iterator i = start; i < end; ++i) {
+        for (Iterator j = i; j > start && compare(*j, *(j - 1)); --j) {
+            swap(*j, *(j - 1));
+        }
+    }
+}
+
+template<typename Iterator>
+void insertion_sort(Iterator start, Iterator end)
+{
+    insertion_sort(start, end, [](auto& a, auto& b) { return a < b; });
+}
+
+template<typename Collection, typename Compare>
+void insertion_sort(Collection& collection, Compare compare)
+{
+    insertion_sort(collection, 0, collection.size() - 1, move(compare));
+}
+
+template<typename Collection>
+void insertion_sort(Collection& collection)
+{
+    insertion_sort(collection, 0, collection.size() - 1, [](auto& a, auto& b) { return a < b; });
+}
```

Comment:
I think these can be collapsed into one with default arguments, though I don't know how well those work with lambdas.

Reply:
I thought about something similar like this, but was a bit hesitant. I could come up with this:
```cpp
// Standard Insertion Sort. NOTE: `end` is inclusive.
template<typename Collection, typename Comparator>
void insertion_sort(
    Collection& collection, int start, int end, Comparator comparator = [](auto& a, auto& b) { return a < b; })
{
    for (int i = start; i <= end; ++i) {
        for (int j = i; j > 0 && comparator(collection[j], collection[j - 1]); --j) {
            swap(collection[j], collection[j - 1]);
        }
    }
}

// NOTE: `end` is exclusive.
template<typename Iterator, typename Comparator>
void insertion_sort(
    Iterator start, Iterator end, Comparator comparator = [](auto& a, auto& b) { return a < b; })
{
    for (Iterator i = start; i < end; ++i) {
        for (Iterator j = i; j > start && comparator(*j, *(j - 1)); --j) {
            swap(*j, *(j - 1));
        }
    }
}

template<typename Collection, typename Comparator>
void insertion_sort(Collection& collection, Comparator comparator)
{
    insertion_sort(collection, 0, collection.size() - 1, move(comparator));
}

template<typename Collection>
void insertion_sort(Collection& collection)
{
    insertion_sort(collection, 0, collection.size() - 1);
}
```
Maybe I misunderstood something here ^^'
This basically collapsed "one" function "away". Did you mean that I could "ignore" the collection and only rely on iterators?

Reply:
This is pretty much exactly what I intended, but you still have one unnecessary function here. The last two can be collapsed as well with the default argument like the first and second.

Reply:
Perfect, I understand now. Thank you! ^^

Reply:
Ouh I think we might have to move the "default argument" to the template, because the compiler is telling me:
```
../../Tests/AK/TestInsertionSort.cpp: In function 'void __test_sorts_ascending()':
../../Tests/AK/TestInsertionSort.cpp:28:23: error: no matching function for call to 'insertion_sort(AK::Vector<int, 10>&)'
   28 |         insertion_sort(ints);
      |         ~~~~~~~~~~~~~~^~~~~~
In file included from ../../Tests/AK/TestInsertionSort.cpp:9:
../.././AK/InsertionSort.h:38:6: note: candidate: 'template<class Collection, class Comparator> void AK::insertion_sort(Collection&, Comparator)'
   38 | void insertion_sort(Collection& collection, Comparator comparator = [](auto& a, auto& b) { return a < b; })
      |      ^~~~~~~~~~~~~~
../.././AK/InsertionSort.h:38:6: note:   template argument deduction/substitution failed:
../../Tests/AK/TestInsertionSort.cpp:28:23: note:   couldn't deduce template parameter 'Comparator'
   28 |         insertion_sort(ints);
      |         ~~~~~~~~~~~~~~^~~~~~

```
Maybe I'm misunderstanding the compile error? In any case, for the sake of it, I'll try it anyway.

Reply:
I think you need better type hints for the comparator, as I was saying in one of my first comments. If you specify that you expect a `Function<bool(T, T)>` where `T` is a different (inferred) type parameter, it should be able to deduce it.

Reply:
I think it doesn't like it being a lambda:
```
../../AK/QuickSort.h:29:23: note:   ‘IDL::generate_overload_arbiter<AK::HashMap<AK::String, AK::Vector<IDL::Function&> >::Entry>(AK::SourceGenerator&, const AK::HashMap<AK::String, AK::Vector<IDL::Function&> >::Entry&, const AK::String&)::<lambda(const auto:38&, const auto:39&)>’ is not derived from ‘AK::Function<int(T, U)>’
   29 |         insertion_sort(col, start, end, less_than);
      |         ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
This is the code right now:
```
template<typename Collection, typename T, typename U>
void insertion_sort(
    Collection& collection, int start, int end,
    Function<int(T, U)> comparator = [](T& a, U& b) { return a < b; })
{
    for (int i = start; i <= end; ++i) {
        for (int j = i; j > 0 && comparator(collection[j], collection[j - 1]); --j) {
            swap(collection[j], collection[j - 1]);
        }
    }
}
```
Is there something I am doing wrong? 🤔 

I also thought about something like this:
```
template<typename Collection, typename Comparator = DefaultComparator>
void insertion_sort(
    Collection& collection, int start, int end, Comparator comparator = Comparator {})
{
    for (int i = start; i <= end; ++i) {
        for (int j = i; j > 0 && comparator(collection[j], collection[j - 1]); --j) {
            swap(collection[j], collection[j - 1]);
        }
    }
}
```
with
```
struct DefaultComparator {
    template<typename T, typename U>
    constexpr int operator()(T& a, U& b)
    {
        return a < b;
    }
};
```